### ISSUE TITLE
LR and dataset sampling

### ISSUE THREAD

**Issue by malicd** [PRBonn/TARL#5]
Hey, I really appreciate your amazing work on this project.
I would be grateful if you could clarify the learning rate and dataset sampling for the semantic segmentation downstream task.

In your paper, you state that you use learning rate of `0.001`. However, this has to be explicitly specified in [segcontrast](https://github.com/PRBonn/segcontrast/blob/a1e961441d23d7dfa1bab3be52de72221114c801/downstream_train.py#L22) framework, otherwise the learning rate of `2.4e-1` will be used. When I specify `--lr 0.001` parameter for downstream fine-tuning of the [provided pretrained model](https://www.ipb.uni-bonn.de/html/projects/tarl/lastepoch199_model_tarl.pt), I obtain the results in 2nd row of the table (I've skipped the experiments noted with `-`).

However, if I omit the `--lr 0.001` argument, or in other words, if I use a learning rate of `2.4e-1`, I am able to get results that are closer to those reported in the paper (last row in the table).

|                             | 0.1%  | 1%    | 10%   | 50%   | 100%  |
|-----------------------------|-------|-------|-------|-------|-------|
| Paper                       | 38.59 | 51.42 | 60.34 | 61.42 | 61.47 |
| Reproduced w/  `--lr 1e-3`  | 27.09 | 32.76 | 37.57 | -     | -     |
| Reproduced w/ `--lr 2.4e-1` | 36.66 | 47.36 | 55.80 | 59.62 | 60.40 |

Could you please clarify the learning rate that you used for the experiments reported in the paper? How does standard deviation looks like between multiple independent runs?
Finally, since the segcontrast framework randomly selects a training subset, do you fix a seed between different runs in order to always train with the same data?

**Comment by nuneslu** [comment]
Hi! Thanks for the interest in our work.

As you mentioned the learning rate used in the paper for the downstream task was indeed `0.001` however we also changed from SGD optimizer to AdamW and used no scheduler (note that the original segcontrast code does use the Cosine Annealing scheduler). I guess the segcontrast repo by default uses SGD but you can change it to the AdamW optimizer in this [line](https://github.com/PRBonn/segcontrast/blob/a1e961441d23d7dfa1bab3be52de72221114c801/trainer/semantic_kitti_trainer.py#L287), replacing from line 287-310 with those 3 lines

```
optimizer = torch.optim.AdamW(optim_params, lr=self.params.lr)

self.optimizer = optimizer
return optimizer
```

Sorry for that, I will add those changes to the segcontrast shortly, but by changing those lines that I mentioned you should be all set to reproduce TARL results. :)

**Comment by malicd** [comment]
Thanks for the fast replay! With AdamW I was able to reproduce the paper results for the 0.1% and 1% data. The other experiments are currently running but I think there should not be any issues.

**Comment by hbzhou530** [comment]
Hi, thanks for your work.
According to the above settings (--lr 1e-3, and change to AdamW), my reproduction accuracy for the semantic segmentation downstream task is unsuccessful(30.74@0.1%, 44.01@1%, 52.60@10%). 
Maybe I missed some details, or environmental problems.
If it is convenient, please help to check my following operations. Thank you so much.

1. conda environment（[TARL_conda.yml.zip](https://github.com/PRBonn/TARL/files/15072776/TARL_conda.yml.zip)）
```
conda create -n TARL python=3.8.16 -y
conda install openblas-devel=0.3.10 -c anaconda
pip3 install -r requirements.txt
pip3 install -U git+https://github.com/NVIDIA/MinkowskiEngine -v --no-deps --install-option="--blas_include_dirs=/home/hi/anaconda3/envs/TARL/include" --install-option="--blas=openblas"
pip3 install -U -e .
```

2. According to [suggestions](https://github.com/PRBonn/TARL/issues/5#issuecomment-1647587812), the learning rate for downstream tasks is modified to 1e-3,  SGD is changed to the AdamW optimizer, and the official pre-training weight is used.

Run scripts using a single GPU by `CUDA_VISIBLE_DEVICES=0`

**Comment by nuneslu** [comment]
Hi @hbzhou530 did you remove the scheduler during fine-tuning? This is also another change, learning rate of 1e-3, AdamW optimizer and **NO** learning rate scheduler.

**Comment by nuneslu** [comment]
To avoid further confusion, I have just created a branch in the segcontrast repo with the hyperparameters and optimizer used for the TARL paper.

**Comment by hbzhou530** [comment]
Thank you for your quick response. After re-checking through your new branch, the accuracy is reproduced (38.07@0.1%). There should be no other problems.  Thx again!

**Comment by nuneslu** [comment]
Glad to hear that! If you have any other questions or problems, feel free to open another issue. :)
